source/MaterialXCore/Element.cpp

```diff
@@ -427,16 +427,24 @@ StringResolverPtr Element::createStringResolver(const string& geom) const
         {
             if (!geomStringsMatch(geom, geomInfo->getActiveGeom()))
                 continue;
-            for (TokenPtr token : geomInfo->getTokens())
-            {
-                string key = "<" + token->getName() + ">";
-                string value = token->getResolvedValueString();
-                resolver->setFilenameSubstitution(key, value);
+            geomInfo->addTokens(resolver);
             }
         }
-    }
+
+    // Add element tokens
+    addTokens(resolver);
 
     return resolver;
+    }
+
+void Element::addTokens(StringResolverPtr& resolver) const
```

Comment:
With Element::addTokens as a virtual method, I'm finding it challenging to reason about the expected behavior for any specific Element subclass, and I wonder if this new feature might be more straightforward as a single, non-virtual method named `StringResolved::addTokenSubstitutions(ConstElementPtr elem)`.  This would make it clearer (to my eyes) what tokens will be added for any given input Element, including traversal to siblings and related interfaces.  What are your thoughts?

Reply:
I think these are some of the points we discussed previously:

1. One of the reasons for this set up was that each element type knows how a token should be added. The alternative would mean the single routine would need to know about derived types and branch on logic.  For example: geomInfo's "<>" and application data "[]" usage would need to be in one piece of logic.

2. Each string resolver usage would need to know how to find which elements to consider for adding tokens and then make this call. As this call would not be "required", it could be omitted and results could be inconsistent. Also having to make an explicit new call would defeat the goal of making token substitution "transparent" to the user -- i.e. current integrations don't need to do any new work.

Reply:
I do agree that it is a bit hard to see the overall logic. Perhaps this could be documented ?